src/gluonts/dataset/split/splitter.py

```diff
@@ -199,60 +192,73 @@ def _train_slice(self, item: TimeSeriesSlice) -> TimeSeriesSlice:
     @abstractmethod
     def _test_slice(
         self, item: TimeSeriesSlice, offset: int = 0
-    ) -> TimeSeriesSlice:
+    ) -> Tuple[TimeSeriesSlice, TimeSeriesSlice]:
         pass
 
-    def _trim_history(self, item: TimeSeriesSlice) -> TimeSeriesSlice:
+    def _trim_history(
+        self, item: Tuple[TimeSeriesSlice, TimeSeriesSlice]
+    ) -> Tuple[TimeSeriesSlice, TimeSeriesSlice]:
         if getattr(self, "max_history") is not None:
-            return item[-getattr(self, "max_history") :]
+            return (
+                item[0][-getattr(self, "max_history") + len(item[1]) :],
+                item[1],
+            )
         else:
             return item
 
-    def split(self, items: List[DataEntry]) -> TrainTestSplit:
-        split = TrainTestSplit()
+    def split(
+        self,
+        items: List[DataEntry],
+    ):
+        test_data = TestDataset(dataset=items, splitter=self)
+        train_data = TrainingDataset(dataset=items, splitter=self)
 
-        for item in map(TimeSeriesSlice.from_data_entry, items):
+        return train_data, test_data
 
+    def _generate_train_slices(self, items: List[DataEntry]):
+        for item in map(TimeSeriesSlice.from_data_entry, items):
             train = self._train_slice(item)
-            test = self._trim_history(self._test_slice(item))
 
-            split._add_train_slice(train)
+            yield train.to_data_entry()
 
-            prediction_length = getattr(self, "prediction_length")
-
-            _check_split_length(train.end, test.end, prediction_length)
-            split._add_test_slice(test)
-
-        return split
-
-    def rolling_split(
+    def _generate_test_slices(
         self,
         items: List[DataEntry],
-        windows: int,
-        distance: Optional[int] = None,
-    ) -> TrainTestSplit:
+        windows: Optional[int],
+        distance: Optional[int],
```

Comment:
the `distance` can't be directly assign with `prediction_length` because when initializing splitter objective, the `prediction_length` is not defined. So I think `distance` remains `Optional[int]` is OK.